### PR TITLE
[Model Zoo] Refactor and Add Utils for Chemistry

### PR DESCRIPTION
**Pull request by mufeili**
## Description

The main purpose of this PR is to support better flexibility for molecular graph featurization and scripts for property prediction.

Previously, we only have a basic method for featurizing atoms in a molecule, i.e. `CanonicalAtomFeaturizer`, which is more like an example for featurizing atoms than something users can easily adapt. I therefore propose to support utilities at different levels:

- Atom/Bond descriptors, which take an RDKit atom/bond object as input and return the descriptor in the form of a list, e.g.
   ```python
   from dgl.data.chem import atom_type_one_hot
   
   mol = Chem.MolFromSmiles('CO')
   atom = mol.GetAtomWithIdx(0)
   atom_type_one_hot(atom)
   # [True, False, ...]
   ```
- An utility function for computing atom/bond descriptors and concatenate the results for a feature.
   E.g.,
   ```python
   featurizer = ConcatFeaturizer(Chem.rdchem.Atom,
                                 [atom_type_one_hot,
                                  atom_degree_one_hot,
                                  atom_implicit_valence_one_hot,
                                  atom_formal_charge,
                                  atom_num_radical_electrons,
                                  atom_hybridization_one_hot,
                                  atom_is_aromatic,
                                  atom_total_num_H_one_hot])
   # [True, False, ..., 0, 0, ...], we can construct a float NumPy array from it later.
   ```
- BaseAtomFeaturizer/BaseBondFeaturizer which take a dictionary of callables for initializing (multiple) features of nodes/edges in a molecular graph. E.g.,
   ```python
   class CanonicalAtomFeaturizer(BaseAtomFeaturizer):
       """A default featurizer for atoms."""
       def __init__(self, atom_data_field='h'):
           super(CanonicalAtomFeaturizer, self).__init__(
                  featurizer_funcs={atom_data_field: ConcatFeaturizer(
                    Chem.rdchem.Atom,
                    [atom_type_one_hot,
                     atom_degree_one_hot,
                     atom_implicit_valence_one_hot,
                     atom_formal_charge,
                     atom_num_radical_electrons,
                     atom_hybridization_one_hot,
                     atom_is_aromatic,
                     atom_total_num_H_one_hot]
                )})
   
   atom_featurizer = CanonicalAtomFeaturizer()
   atom_featurizer(mol)
   """
   {'h': tensor([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 0.],
                 [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0.]])}
   """
   ```

Also copy @MilkshakeForReal , who might be interested to have some discussions.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by MilkshakeForReal** [comment]
Thanks for informing me. The refactoring is nice and clean. Moreover, I want to emphasize the importance of bond features especially when using MPNN model. It would be even better if we could add `CanonicalBondFeaturizer`, e.g. the bond featurizer in `data/chem/alchemy`, which is relatively a simply one and could be directly implemented. To enrich the bond featurizer if anyone want, check bellow:
featurest

example code to create above rdkit features:
https://iwatobipen.wordpress.com/2019/04/05/make-graph-convolution-model-with-geometric-deep-learning-extension-library-for-pytorch-rdkit-chemoinformatics-pytorch/
https://github.com/yongqyu/MolGAN-pytorch
https://deepchem.io/docs/_modules/deepchem/feat/graph_features.html#bond_features
Another interesting thing is that maybe we could add `g.is_in_ring(eids)` as a more general api?

**Comment by MilkshakeForReal** [comment]
Awkward to find out that the work had been done, which made my work redundant lol. But that’s what I admire about dgl most. 👍

**Comment by mufeili** [comment]
@MilkshakeForReal Thanks for the kind words.

1. `g.is_in_ring(eids/nids)`
    I think currently we can just use the RDKit utilities for ring membership testing. Either we directly treat the ring memberships as a node/edge feature or we can keep the SMILES and compute ring membership when needed.
2. Node/Edge Featurization for non-bidirected graphs
    I'm a bit tentative about that. The reason is that node/edge featurization is not very well defined when the graphs are non-bidirected. There can be various possibilities for virtual nodes/edge.  I can probably add some existing methods from relatively impactful work if you know any.

File: python/dgl/data/chem/utils.py [hunk 1]
```diff
@@ -10,42 +12,583 @@
...
-    All atom featurizers that map a molecule to atom features should subclass it.
-    All subclasses should overwrite ``_featurize_atom``, which featurizes a single
-    atom and ``__call__``, which featurizes all atoms in a molecule.
+    Loop over all atoms in a molecule and featurize them with the ``featurizer_funcs``.
+
+    **We assume the resulting DGLGraph will not contain any virtual nodes.**
+
+    Parameters
+    ----------
+    featurizer_funcs : dict
```
Comment (VoVAllen) [inline python/dgl/data/chem/utils.py]: Why does this need to be a dict?

File: python/dgl/data/chem/utils.py [hunk 2]
```diff
@@ -10,42 +12,583 @@
...
+        The type of objects that we want to featurize. E.g. ``rdkit.Chem.rdchem.Atom``
+        or ``rdkit.Chem.rdchem.Bond``.
+    func_list : list
+        List of functions for computing molecular descriptors from objects of type
+        ``data_type``. Each function is of signature
+        ``func(data_type) -> list of float or bool or int``. The resulting order of
+        the features will follow that of the functions in the list.
+    """
+
+    def __init__(self, data_type, func_list):
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/chem/utils.py]: Strictly speaking, this can be used for any type of the data like atom, bond or molecule. To declare the type of the target for featurization is just for a data type checking during featurization.

File: examples/pytorch/model_zoo/chem/property_prediction/utils.py [hunk 3]
```diff
@@ -68,6 +78,68 @@ def roc_auc_averaged_over_tasks(self):
             total_score += roc_auc_score(task_y_true, task_y_pred)
         return total_score / n_tasks
 
+    def l1_loss_averaged_over_tasks(self):
```
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/pytorch/model_zoo/chem/property_prediction/utils.py]: How about making this behave like the loss functions in PyTorch, i.e. `reduction` can be `none`, `mean` or `sum`.

File: python/dgl/data/chem/utils.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/data/chem/utils.py]: I think this is the easiest thing we can do when we want to construct multiple features at one pass.

File: python/dgl/data/chem/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/data/chem/utils.py]: It's user's duty to ensure the function and the variable's type are matched in Python. For dynamic language, this is a bit weird. Another reason is `rdkit.Chem.rdchem.Atom` is too long to memorize. It's not ideal for user to pass this in.

File: python/dgl/data/chem/utils.py (hunk 1 above)
Comment (VoVAllen) [reply python/dgl/data/chem/utils.py]: I don't like this but it's fine. Could you add an example at the doc?

**Comment by VoVAllen** [comment]
I doubt you were a Java programmer before, otherwise you won't like typing and inheritance so much (and really long function names).  😄

File: python/dgl/data/chem/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/data/chem/utils.py]: Done.

File: examples/pytorch/model_zoo/chem/property_prediction/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/pytorch/model_zoo/chem/property_prediction/utils.py]: I now use reduction only for "l1_loss" to decide whether to sum or average over all labeled data points for each task. The returned result is always a list of floats for all tasks.

File: python/dgl/data/chem/utils.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/data/chem/utils.py]: Done.

**Comment by mufeili** [comment]
> I doubt you were a Java programmer before, otherwise you won't like typing and inheritance so much (and really long function names). 😄

No, I'm a believer of Murphy's law :smiling_imp:
